File: problem_19/counting_sundays_test.rb
require 'minitest/autorun'
require 'minitest/pride'

class CountingSundaysTest < Minitest::Test
  def first_month_sundays(start_year, end_year)
    count = 0

    (start_year..end_year).each do |year|
      day = first_sunday(year)
      until day > 336
        if leap_year?(year)
          count += 1 if LEAP_FIRSTS_OF_THE_MONTH.include?(day)
        else
          count += 1 if NON_LEAP_FIRSTS_OF_THE_MONTH.include?(day)
        end
        day += 7
      end
    end

    count
  end

  def first_sunday(year)
    day = 7
    total_days = 0
    leap_count = 0
    non_leap_count = 0

    (1900..(year - 1)).each do |yr|
      if leap_year?(yr)
        leap_count += 1
      else
        non_leap_count += 1
      end

      total_days = (366 * leap_count) + (365 * non_leap_count)
      day += 7 until day > total_days
    end

    day - total_days
  end

  def leap_year?(year)
    if year % 100 != 0
      year % 4 == 0
    else
      year % 400 == 0
    end
  end

  #----------------- Tests -----------------#

  def test_first_month_sundays
    assert_equal 3,   first_month_sundays(1901, 1902)
    assert_equal 171, first_month_sundays(1901, 2000)
  end

  def test_first_sunday
    assert_equal 6, first_sunday(1901)
    assert_equal 5, first_sunday(1902)
    assert_equal 3, first_sunday(2010)
    assert_equal 5, first_sunday(2020)
  end

  def test_leap_year
    assert_equal true,  leap_year?(1992)
    assert_equal true,  leap_year?(2400)
    assert_equal false, leap_year?(1900)
    assert_equal false, leap_year?(1997)
  end

  #----------------- Constants -----------------#

  NON_LEAP_FIRSTS_OF_THE_MONTH = [1, 32, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335]
  LEAP_FIRSTS_OF_THE_MONTH =     [1, 32, 61, 92, 122, 153, 183, 214, 245, 275, 306, 336]
end

File: problem_19/counting-sundays.test.js
//----------------- Constants -----------------//

const nonLeapFirsts = [1, 32, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335];
const leapFirsts = [1, 32, 61, 92, 122, 153, 183, 214, 245, 275, 306, 336];

//----------------- Functions -----------------//

function firstOfTheMonthSundayCount(startYear, endYear) {
  let count = 0;
  let day = 7;

  for (year = 1900; year <= endYear; year++) {
    while (day <= 336) {
      if (leapYear(year)) {
        if (year >= startYear && leapFirsts.includes(day)) {
          count += 1;
        }
      } else {
        if (year >= startYear && nonLeapFirsts.includes(day)) {
          count += 1;
        }
      }
      day += 7;
    }

    if (leapYear(year)) {
      day = ((day + 7 * 5) % 366) % 7;
    } else {
      day = ((day + 7 * 5) % 365) % 7;
    }
  }

  return count;
}

function leapYear(year) {
  if (year % 100 !== 0) {
    return year % 4 === 0;
  } else {
    return year % 400 === 0;
  }
}

//----------------- Tests -----------------//

test("firstOfTheMonthSundayCount", () => {
  expect(firstOfTheMonthSundayCount(1900, 1901)).toBe(4);
  expect(firstOfTheMonthSundayCount(1901, 1902)).toBe(3);
  expect(firstOfTheMonthSundayCount(1901, 2000)).toBe(171);
});

test("leapYear", () => {
  expect(leapYear(1992)).toBe(true);
  expect(leapYear(2400)).toBe(true);
  expect(leapYear(1900)).toBe(false);
  expect(leapYear(1997)).toBe(false);
});
